Merge work from MCP feature branch

Azquelt opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Add `mcp-1.0` to allow tools to be written an exposed over the [MCP protocol][mcp] from liberty.

All of the work here has been reviewed when it was added to the `feature-mcp-branch`.

[mcp]: https://modelcontextprotocol.io/specification/2025-06-18

Azquelt committed:
Initial MCP prototype

- Only supports calling a tool
- Only supports strings as parameters
- Doesn't support sessions or requests yet
- No authentication
- Almost no tests

Azquelt committed:
Add MCP FAT project

Doesn't test much yet

habiblawal1 committed:
fix: check presence of required mcp-server request params

habiblawal1 committed:
Update parsing of tool/call parameters

Put the parameters into their own object rather than extending
McpRequest.

Azquelt committed:
Add support for the Initialization message

IssacAX123 committed:
Adds strict and leniant JSONAssert tests

IssacAX123 committed:
deletes JSONTest and fixes logical errors in ToolTest

Azquelt committed:
Merge pull request #32316 from IssacAX123/32294-jsonassert

32294 - Adds example strict and lenient JSONAssert tests

habiblawal1 committed:
docs: add javadoc to McpRequest.java

Azquelt committed:
Merge pull request #32317 from habiblawal1/issue/32292-robust-mcp-request-parsing

More Robust Parsing of MCP Tool Request Parameters

habiblawal1 committed:
feat: support string and number types for jsonrpc request id

habiblawal1 committed:
test: FAT and unit tests for string and number jsonrpc request ids

devalibm committed:
mcp-server : HTTP validations

devalibm committed:
mcp-server : HTTP validations

IssacAX123 committed:
adds jsonrpc error code exceptions class

Azquelt committed:
Merge pull request #32326 from habiblawal1/issue/32324-correct-id-support

Support String and Number JSON-RPC request ids

IssacAX123 committed:
Adds JSONRPC exception handling to MCPServlet

IssacAX123 committed:
Adds JSONRPC exception handling to MCPServlet

martindrozdz committed:
MCP List Tools functionality

devalibm committed:
test : For accept headers and header validations

devalibm committed:
test : For accept headers and header validations

devalibm committed:
Review: resolves code review issue

devalibm committed:
HttpServlet : adds new field

devalibm committed:
HttpServlet : adds new field

devalibm committed:
wip

devalibm committed:
fix: correct logic for checking mime type

devalibm committed:
fix: correct logic for checking mime type

martindrozdz committed:
MCPListToolsUpdate

Azquelt committed:
revert changes to debug launch files

martindrozdz committed:
MCPListToolsReviewUpdate

martindrozdz committed:
Merge pull request #32355 from Azquelt/revert-debug-file-changes

revert changes to debug launch files

IssacAX123 committed:
adds matchers for exception testing

devalibm committed:
fat-test : postJsonRpc

Azquelt committed:
Merge pull request #32358 from Azquelt/mcp-init

Add support for the Initialization message

Azquelt committed:
Merge pull request #32356 from martindrozdz/feature-32291

MCPListToolsReviewUpdate

devalibm committed:
amendments : removes unnecessary path from classpath

devalibm committed:
Review Changes

martindrozdz committed:
MCPListToolsFATTest

devalibm committed:
Review Changes : also added an response LifeCycleTest

martindrozdz committed:
MCPListToolsAddedFATTest

martindrozdz committed:
MCPListTools Added Tool and expected output

martindrozdz committed:
MCPListTools PR review changes

martindrozdz committed:
MCPListTools PR review changes

Azquelt committed:
Merge pull request #32371 from martindrozdz/feature-32291

Feature 32291 - MCP List Tools

devalibm committed:
Fat test : adds util method

devalibm committed:
Fat test : updates the reponse for rebased code

martindrozdz committed:
MCPListTools Added Tool Title and arg description

IssacAX123 committed:
Adds FAT test for exception handling

devalibm committed:
Fat Test : removes falseCallMCP tests

IssacAX123 committed:
adds FDCC ignore

devalibm committed:
Fat Test : removes white spaces

Azquelt committed:
Merge pull request #32328 from devalibm/mcp-prototype

Basic Validations for HTTP requests

IssacAX123 committed:
rebase fix

ttmunyunguma committed:
add type conversion

ttmunyunguma committed:
Merge remote-tracking branch 'upstream/feature-mcp-server' into feature-mcp-server

# Conflicts:
#	dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolTest.java

ttmunyunguma committed:
add type conversion

IssacAX123 committed:
Convrets to use McpResponse instead of ToolResponse

ttmunyunguma committed:
revert type conversion

IssacAX123 committed:
PR fix changes code to data

habiblawal1 committed:
refactor: move servlet code for sending response to McpTransport

IssacAX123 committed:
PR fix extracts result

IssacAX123 committed:
PR fixex

IssacAX123 committed:
PR fixes

IssacAX123 committed:
Merge pull request #32425 from IssacAX123/32296-graceful-exceptions

32296 graceful exceptions

ttmunyunguma committed:
add tool tests

ttmunyunguma committed:
Merge branch 'OpenLiberty:feature-mcp-server' into feature-mcp-server

ttmunyunguma committed:
update tool response to return generic object

ttmunyunguma committed:
Merge remote-tracking branch 'origin/feature-mcp-server' into feature-mcp-server

# Conflicts:
#	dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolResponseResult.java

ttmunyunguma committed:
test for boolean input args

IssacAX123 committed:
Adds tool call error responses

habiblawal1 committed:
Merge pull request #32456 from IssacAX123/32373-tool-error-handling

Adds tool call error responses

ttmunyunguma committed:
Merge branch 'feature-mcp-server' of github.com:OpenLiberty/open-liberty into feature-mcp-server

# Conflicts:
#	dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolResponseResult.java
#	dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolTest.java

ttmunyunguma committed:
fix test with new impl

IssacAX123 committed:
adds specefic argumenst in error data message

habiblawal1 committed:
Merge pull request #32451 from ttmunyunguma/feature-mcp-server

Feature mcp server - Fixes #32375 Support more argument and response types for tools

IssacAX123 committed:
replaces apche set with java utils

IssacAX123 committed:
sends null data if error check fails

ttmunyunguma committed:
Handle initialized notification from the client

habiblawal1 committed:
Merge pull request #32476 from IssacAX123/32458-detailed-argument-error

#32458 Detailed Mismatch Argument Error

IssacAX123 committed:
optimal args set complement

Azquelt committed:
Merge pull request #32491 from IssacAX123/32458-detailed-argument-error

optimal args set complement

ttmunyunguma committed:
log response body when exception is thrown

devalibm committed:
Correct bean lifecycle

Make sure we destroy dependent beans after running tools.

devalibm committed:
Merge pull request #32475 from devalibm/bean-lifecycle

Bean lifecycle

ttmunyunguma committed:
fix merge review comments

ttmunyunguma committed:
Merge pull request #32511 from ttmunyunguma/feature-mcp-server-32431

Handle initialized notification from the client

devalibm committed:
Respond to GET request with 405

Azquelt committed:
Merge pull request #32535 from devalibm/reject-get-method

Respond to GET request with 405

devalibm committed:
Check MCP Protocol Version

martindrozdz committed:
Tool name & tile annotations changes

Azquelt committed:
Add the rest of the API for tools

Imported from quarkus-mcp-server

This gives us a more complete API for tools.

martindrozdz committed:
Added Tests

martindrozdz committed:
Updates for new ToolMetadata fields

martindrozdz committed:
Reverted server.xml and bootstrap logging

martindrozdz committed:
Removed generated bootstrap file

ttmunyunguma committed:
add a connection tracker to mcp tool calls and register cancellations

Azquelt committed:
Merge pull request #32551 from Azquelt/mcp-tools-api

Add the rest of the API for tools

devalibm committed:
Fix : Resolves code review conversations

martindrozdz committed:
Fixed Tests

martindrozdz committed:
Added Final and Blank comparison

devalibm committed:
Test : Adds new test to FAT bucket

devalibm committed:
Fix : apply correct case to variables

devalibm committed:
Merge pull request #32549 from devalibm/mcp-version-header

Check MCP Protocol Version

devalibm committed:
FAT Test : Adds testRejectsUnsupportedProtocolVersion

Azquelt committed:
Merge pull request #32584 from devalibm/mcp-version-header

FAT Test : Adds testRejectsUnsupportedProtocolVersion

martindrozdz committed:
Moved ToolMetaData creation for tests to ToolMetadataUtil

IssacAX123 committed:
adds test for tool not found

martindrozdz committed:
Added Wrapper types for ToolList

martindrozdz committed:
Merge pull request #32382 from martindrozdz/feature-32372

MCPListTools Added Tool Title and arg description

IssacAX123 committed:
Merge pull request #32588 from IssacAX123/32555-jsonrpc-missing-method-test

32555 jsonrpc missing method test

IssacAX123 committed:
adds ping support

martindrozdz committed:
Added Fat Tests for Primitives and Objects

devalibm committed:
Adds Fix : Support tool description Annotations

habiblawal1 committed:
Merge pull request #32560 from habiblawal1/issue/32374-http-abstraction

Abstract HTTP Response methods to McpTransport.java

martindrozdz committed:
Removed Tool Registry

Azquelt committed:
Merge pull request #32593 from IssacAX123/32558-ping-method

adds ping support

martindrozdz committed:
Merge pull request #32590 from martindrozdz/feature-32573

Feature 32573 - Testing primitives and wrapper objects for MCP Tool List

martindrozdz committed:
Consolidated JsonB to setup()

habiblawal1 committed:
refractor: set header and contentType in HttpResponseException class

martindrozdz committed:
Added primitive and Wrapper tests to tools/call

martindrozdz committed:
Removed Debug Sysout

martindrozdz committed:
Added decimals for floats and doubles types

habiblawal1 committed:
refractor: use lambda for iterating over headers

martindrozdz committed:
Made Expected number response to be a string according to spec

martindrozdz committed:
Test cases for empty & null Tool info

Azquelt committed:
Fix wrong Content class import

Azquelt committed:
Add Tool Exception API classes

From quarkus-mcp-server

martindrozdz committed:
Tool Return type is now always String

martindrozdz committed:
Objects.toString() for transport response

martindrozdz committed:
Merge pull request #32610 from martindrozdz/feature-32573

Feature 32573 - Tests added for Wrapper and primitive types (Tool call & list)

martindrozdz committed:
Merge pull request #32621 from martindrozdz/feature-32372

Feature 32372 - Test cases for empty & null Tool info

habiblawal1 committed:
refractor: use mehtod chaining for adding headers to a HttpResponseException

Azquelt committed:
Merge pull request #32622 from Azquelt/mcp-tool-api

Add Tool Exception API classes

habiblawal1 committed:
refractor: return empty map if no headers are set in the HttpResponseException class

devalibm committed:
Fix : Adresses the review changes

habiblawal1 committed:
fix: set content type to text/plain

habiblawal1 committed:
Merge pull request #32606 from habiblawal1/issue/32604-transport-http-exception

Set Header and ContentType in HttpResponseException class

devalibm committed:
fix : Further changes

devalibm committed:
Fix : IgnoredEcho Tool

devalibm committed:
Fix : Fixes emoty tool title

martindrozdz committed:
Fixed name

devalibm committed:
Refactor : final fixes

devalibm committed:
Merge pull request #32605 from devalibm/tool-annotations

Adds feat : Support tool description Annotations

ttmunyunguma committed:
fix cancellation issues of noclassdefexception

ttmunyunguma committed:
use callable future in FAT test and cleanup

martindrozdz committed:
Duplicate Tool Error Management

devalibm committed:
Feat : Supports more Contents for tools

devalibm committed:
Fix : Review changes

Azquelt committed:
Merge pull request #32690 from devalibm/content-objects

Supports more Contents for tool

devalibm committed:
feat : Support for response Content

devalibm committed:
feat : Adds new tools and tests

devalibm committed:
Fix : Review changes

devalibm committed:
Fix : Further review changes

devalibm committed:
Merge pull request #32694 from devalibm/content-objects

feat : Support for response Content

Azquelt commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=713228b8-db4d-43c5-ab48-99ea82660a46) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

devalibm committed:
Refactor : Cleans ToolTest

devalibm committed:
Fix : Review changes

devalibm committed:
Merge pull request #32698 from devalibm/amends-tooltest

Refactor : Cleans ToolTest

habiblawal1 committed:
Enable support for tools to return complex objects (#32680)

* feat: enable support to return complex objects

* test: tests for tools responding lists and arrays

* feat: change structuredContent to be chosen when declaring Tool annotation

* fix: do not return structured content if tool returns string

* docs: update Tool annotation javadoc to include structuredContent and OutputSchema

* chore: add correct copyright headers

---------

Co-authored-by: Habib Lawal <Habib.Lawal@ibm.com>

ttmunyunguma committed:
Fixing review comments

jhanders34 commented on dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolDescription.java:
```diff
@@ -0,0 +1,131 @@
+
+        InputSchemaPrimitive tempSchemaPrimitive;
+
+        boolean isJsonNumber = type.equals(long.class) || type.equals(double.class) || type.equals(byte.class) || type.equals(float.class) || type.equals(short.class)
```
Probably would be easier to do with a static Set that you just do a contains passing in type instead of all of these ORs and equals.  Or having a map that returns an impl of a Converter interface that returns the values.  If one isn't in the map, then it just does the final else.

Azquelt replied:
This is all early prototype code. I expect this part will end up being replaced in #32572 when we extend schema support.

jhanders34 commented on dev/io.openliberty.mcp.internal_fat/bnd.bnd:
```diff
@@ -0,0 +1,33 @@
+	com.ibm.ws.logging;version=latest,\
+	com.ibm.websphere.appserver.spi.logging,\
+	com.ibm.ws.logging.core,\
+	io.openliberty.jakarta.annotation.3.0
```
```suggestion
	io.openliberty.jakarta.annotation.2.1
```
EE 10 version is 2.1.

jhanders34 commented on dev/io.openliberty.mcp.internal_fat/build.gradle:
```diff
@@ -0,0 +1,38 @@
+ *******************************************************************************/
+apply from: '../wlp-gradle/subprojects/repos.gradle'
+
+configurations {
+  json
+}
+
+configurations.json {
+  transitive = false
+}
+
+dependencies {
+  json 'org.json:json:20080701',
+    'org.skyscreamer:jsonassert:2.0-rc1'
+}
+
+
+task addJson(type: Copy) {
+  from configurations.json
+  into "${buildDir}/autoFVT/lib"
+}
+
+addRequiredLibraries {
+  dependsOn addJson
+}
+
+addRequiredLibraries.dependsOn addJakartaTransformer
```
```suggestion
configurations.requiredLibs {
  transitive = false
}

dependencies {
  requiredLibs 'org.json:json:20080701',
               'org.skyscreamer:jsonassert:2.0-rc1'
}
```

jhanders34 commented on dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/FATSuite.java:
```diff
@@ -0,0 +1,36 @@
+                ToolTest.class
+
+})
+public class FATSuite {
```
Do we need to do repeats for EE 10 and 11?  Not sure if repeat rules handle `platform` yet for versionless features.

Azquelt replied:
We will eventually, but not yet

jhanders34 commented on dev/io.openliberty.mcp/src/io/openliberty/mcp/annotations/OutputSchemaGenerator.java:
```diff
@@ -0,0 +1,29 @@
+/*******************************************************************************
```
Should these also have EPL license?

Azquelt replied:
I didn't think so but I'm happy to be told what it should say. Who would I need to ask about this?

jhanders34 commented on dev/io.openliberty.mcp/src/io/openliberty/mcp/annotations/WrapBusinessError.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Based on https://github.com/quarkiverse/quarkus-mcp-server/blob/main/core/runtime/src/main/java/io/quarkiverse/mcp/server/Tool.java
```
Should this say WrapBusinessError instead of Tool?

Azquelt replied:
yes

jhanders34 commented on dev/io.openliberty.mcp/src/io/openliberty/mcp/content/Content.java:
```diff
@@ -0,0 +1,97 @@
+    enum Type {
+        TEXT,
+        IMAGE,
+        RESOURCE,
+        AUDIO,
+        RESOURCE_LINK
+    }
```
Should `RESOURCE` and `RESOURCE_LINK` be removed?

Azquelt replied:
Doesn't really matter, we do plan to support them eventually.

jhanders34 commented on dev/com.ibm.websphere.appserver.features/visibility/public/mcp-1.0/io.openliberty.mcp-1.0.feature:
```diff
@@ -0,0 +1,23 @@
+IBM-App-ForceRestart: install, \
+ uninstall
+IBM-API-Package: \
+ io.openliberty.mcp, \
+ io.openliberty.mcp.annotations, \
+ io.openliberty.mcp.content, \
+ io.openliberty.mcp.encoders, \
+ io.openliberty.mcp.messaging, \
```
```suggestion
 io.openliberty.mcp.annotations, \
 io.openliberty.mcp.content, \
 io.openliberty.mcp.encoders, \
```
Those two packages don't seem to exist.

Azquelt replied:
Messaging will be added in #32521, I'll remove the other one.

jhanders34 commented on dev/com.ibm.websphere.appserver.features/visibility/public/mcp-1.0/io.openliberty.mcp-1.0.feature:
```diff
@@ -0,0 +1,23 @@
+ io.openliberty.mcp.messaging, \
+ io.openliberty.mcp.meta, \
+ io.openliberty.mcp.tools
+IBM-ShortName: mcp-1.0
```
`build.featureStart.base/resources/com/ibm/ws/test/featurestart/features/feature-levels.properties` will need to be updated to have `mcp-1.0=17` added to it.

Azquelt replied:
Thanks, I wasn't aware of that file.

ttmunyunguma committed:
Use collections

martindrozdz committed:
Added 2nd app with duplicate methods

martindrozdz committed:
Added 2nd app with duplicate methods 2/2

martindrozdz committed:
Updated added Tests

martindrozdz committed:
Added DeploymentProblemTest.class

martindrozdz committed:
Removed logging traceSpecification

martindrozdz committed:
Added testToolArgNotRequired method

ttmunyunguma committed:
Fix review comments

Azquelt commented on dev/fattest.simplicity/src/componenttest/topology/utils/HttpRequest.java:
```diff
@@ -274,6 +282,10 @@ public int getResponseCode() {
         return responseCode;
     }
 
+    public Map<String, String> getResponseHeaders() {
+        return responseHeaders;
+    }
```
This causes a compile failure in `com.ibm.ws.ui_rest_fat` because they extend `HttpRequest` and add a method `public Map<String,List<String>> getResponseHeaders()`

Azquelt commented on dev/fattest.simplicity/src/componenttest/topology/utils/HttpRequest.java:
```diff
@@ -243,7 +249,9 @@ public <T> T run(Class<T> type) throws Exception {
-                throw new Exception("Unexpected response: " + responseCode);
+                String responseBody = request.getResponseBodyAsString();
+                Log.info(c, "run", "Response body: " + responseBody);
+                throw new Exception("Unexpected response: " + responseCode + "\nResponse Body: " + responseBody);
```
We should trim the response body here. If we get a bad status code, it's likely either a short error message or a huge HTML error page.

martindrozdz committed:
Added review changes

Azquelt committed:
Add mcp-1.0 to the feature levels list

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
MCP: Remove non-existent packages from API list

messaging remains because that's currently in development

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
Fix MCP FAT dependencies

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
MCP: Fix WrapBusinessError copyright header

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

Azquelt committed:
Fix up HttpRequest changes made for MCP

- remove getResponseHeaders - it wasn't used and conflicts with the
  ui.rest fat which extends this class
- truncate an error response when logging it (as we do for successful
  responses)

Azquelt committed:
MCP: add jsonb feature dependency

Azquelt committed:
Repeat MCP FAT on EE11

martindrozdz committed:
Merge pull request #32656 from martindrozdz/feature-32620

feature-32620 -- DuplicateToolError Implementation with new FAT Tests

martindrozdz committed:
Added changes for rebase

Azquelt committed:
Merge pull request #32732 from Azquelt/mcp-integration-review

MCP Feature integration review fixes

Azquelt commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=da9c88e1-ac13-433b-9852-fef4999e6b4e) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

martindrozdz committed:
Merge pull request #32719 from martindrozdz/ToolArgRequired

Tool arg required - #32721

Azquelt committed:
MCP: remove trailing space from bnd.bnd

Azquelt committed:
MCP: Fix MessageParsingTest

Pass correct arguments when creating the test ArgumentMetadata

Azquelt committed:
Merge pull request #32748 from Azquelt/fix-message-parsing-test

MCP: Fix MessageParsingTest

ttmunyunguma committed:
Merge branch 'feature-mcp-server' into feature-mcp-server-32521

# Conflicts:
#	dev/io.openliberty.mcp.internal/test/io/openliberty/mcp/internal/test/MessageParsingTest.java
#	dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/FATSuite.java
#	dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/basicToolApp/BasicTools.java

ttmunyunguma committed:
fixing review comments

Azquelt committed:
Update EE9 test: mcp-1.0 is not compatible

habiblawal1 committed:
feat: build mcp jar in dev/api/ibm (#32749)

Azquelt committed:
Merge pull request #32750 from Azquelt/mcp-ee9-test-fix

Update EE9 test: mcp-1.0 is not compatible

Azquelt committed:
Merge pull request #32752 from Azquelt/mcp-bnd-fix

MCP: remove trailing space from bnd.bnd

Azquelt committed:
MCP: Feature description review comments

ttmunyunguma committed:
Merge pull request #32674 from ttmunyunguma/feature-mcp-server-32521

Allow users to cancel tool calls #32521

Azquelt commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=9283de68-62a9-477d-bfad-98b6080ac0d3) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

Azquelt committed:
MCP: Create special arguments in one place

Previously we had a two-step process for creating the values for special
arguments before calling a tool.

After parsing the arguments from the request, we would then insert a
placeholder for every special argument. Later, in McpServlet.callTool,
we would have those placeholders replaced by the actual values.

Instead of replacing placeholders, we can use the tool metadata to find
which parameters need special arguments and what types they should be so
that all the processing can be done in one place.

Azquelt committed:
Merge pull request #32757 from Azquelt/mcp-move-special-argument-creation

MCP: Create special arguments in one place

Azquelt committed:
Rename the mcp server feature

Rename from mcp-1.0 to mcpServer-1.0 since we may one day want to
provide a client MCP API.

Azquelt commented:
In addition to the changes from the review comments in this PR, there were a few more changes I've had to make to get the build to pass.

- #32752
- #32750

A message reviewer also requested some changes which are being done here
- #32753 

I've also realised that while making a PR off the feature branch itself made sense, it does mean that more things are constantly being delivered to it. Once I have a good personal build, I'll make a copy of the feature branch and do a PR from the copy so that it won't receive any more changes while the build is running.

Azquelt commented:
Failures in the mcp FAT project show the cancellation test is unreliable. It's possible for the cancellation to arrive and be processed before the tool call has started - despite a 1 second wait between starting the tool and trying to cancel it.

Azquelt committed:
MCP CancellationTest: use separate ids

Use a different id for each test (as a client would be expected to) so
that the tests can't interfere. Otherwise if first test fails, second
test can fail because the first request is still ongoing.

Azquelt committed:
MCP: move cancellation test tools to their own app

Azquelt committed:
MCP CancellationTool: await tool started

Wait for the tool to start running before trying to cancel it.

Azquelt committed:
MCP CancellationTest: improve logging

Azquelt committed:
Fix mcpServer feature files

- Use private features for tolerated dependencies
- Add types to APIs
- Add dependency on noShip feature

Azquelt committed:
Merge pull request #32753 from Azquelt/mcp-update-feature-description

MCP: Feature description review comments

Azquelt committed:
Merge pull request #32762 from Azquelt/mcp-rename-feature

Rename the mcp server feature to mcpServer-1.0

Azquelt committed:
Merge pull request #32772 from Azquelt/mcp-cancellation-test-reliability

Mcp cancellation test reliability

Azquelt commented:
The build is now passing (see #32773). I'm closing this PR in favour of #32776, which has the same content, but is based off a different branch so that any new merges into the `feature-mcp-server` branch don't get automatically added to the PR.
